crates/engine/query-solver/src/solve/steiner_tree/shortest_path/tree.rs

```diff
@@ -0,0 +1,125 @@
+use petgraph::{
+    graph::GraphIndex,
+    visit::{EdgeRef, GraphBase, NodeCount, NodeIndexable},
+};
+
+use crate::{solve::steiner_tree::SteinerGraph, Cost};
+
+use super::{Reusable, SteinerContext};
+
+#[derive(Clone)]
+pub(super) struct ShortestPathSteinerTree<G: GraphBase<NodeId: Copy, EdgeId: Copy>> {
```

Comment:
Mostly copy-pasted, split it to a new file + more generics for a bit more safety in the indices